Add donation goals

begedin opened the pull request:
# What's in this PR?
## `donation-goal``

A component for viewing a donation goal. Can send an `edit` action to the parent, indicating that the "Edit" link rendered by the component was clicked.

``` handlebars
{{donation-goal 
  donationGoal=donationGoal 
  edit=(action externalEditHandler donationGoal)}}
```
## `donation-goal-edit`

A component for editing a donation goal

The component is not added anywhere into the app, but it's well tested with integration tests. I tried to finally use the "proper" data-down-actions-up approach here (or I hope I did, someone with more ember knowledge can correct me on it.

The component has two actions, `save` and `cancel`. In order to use it correctly, we do the following:

``` handlebars
{{donation-goal-edit
  cancel=(action externalCancelHandler)
  donationGoal=donationGoal
  save=(action externalSaveHandler donationGoal)}}
```
## `donation-goals`

A component for displaying a list of donation goals. Each can be edited/cancelled/saved individually

``` handlebars
{{donation-goals
  add=(action 'addDonationGoal')
  cancel=(action 'cancelDonationGoal')
  edit=(action 'editDonationGoal')
  project=project
  save=(action 'saveDonationGoal')}}
```

The action handlers will both receive the donation goal which is being saved/cancelled.
## Route behavior

Initially, all of the project's donation goals are listed. If there are none, a new record is initialized and added to the list. 
- Saving a new or existing records commits those changes to the API.
- Cancelling a new record removes that record from the list.
- Cancelling an existing record simply puts it back to display mode.
## TODO
- [x] `donation-goal-edit` component
- [x] `donation-goal` component
- [x] `donation-goals` component
- [x] `donation-goal` route
- [x] UI details (texts, etc.)
- [x] Hide the `edit` link and `add new buttons` for unauthorized users
- [x] add missing route and model unit tests (basic, but did not create them yet).
- [x] rewrite `donation-goal-edit` to bind to the donationGoal, not individual properties. Should then set initial values of individual properties on init. More explicit that way, but easier to use.
* [x] Merge #654 
* [x] Add it to the `donation-goal-edit` component here.
* [x] Update tests
* [x] Add project ability tests

## References

Progress on  #470
Progress on: #469

sbatson5 commented on app/templates/components/edit-donation-goal.hbs:
```diff
@@ -0,0 +1,16 @@
+  <span class="period">per month</span>
+</div>
+<div class="input-group">
+  {{textarea name="description" value=description placeholder="Tell your donors what this goal will allow you to do. Be specific, but make it interesting!"}}
```
Thoughts on just breaking this onto multiple lines just for readability?

```
{{textarea
  name="description"
  value=description
  placeholder="Tell your donors what this goal will allow you to do. Be specific, but make it interesting!"}}
```

JoshSmith replied:
Agree here.

JoshSmith commented on app/components/edit-donation-goal.js:
```diff
@@ -0,0 +1,44 @@
+    amount=donationGoal.amount
+    description=donationGoal.description
+    save=(action externalSaveHandler donationGoal)
+    cancel=(action externalCancelHandler)}}
```
Should we alphabetize here?

JoshSmith commented on tests/integration/components/edit-donation-goal-test.js:
```diff
@@ -0,0 +1,65 @@
+  this.render(hbs`{{edit-donation-goal isNew=false amount=1 description='Test' cancel=(action cancelHandler donationGoal)}}`);
+
+  this.$('.cancel').click();
+});
```
Can we use page objects here?

begedin replied:
It's an integration test. Are we using them in other integration tests?

JoshSmith commented on app/styles/components/edit-donation-goal.scss:
```diff
@@ -0,0 +1,26 @@
+    float: left;
+    margin: 0 0 10px 0;
+
+    .currency, input[type=number], .period {
```
Yeah, these will need changed.

JoshSmith commented on app/components/edit-donation-goal.js:
```diff
@@ -0,0 +1,44 @@
+
+    cancel() {
+      this.sendAction('cancel');
+    }
```
Looks like these are not alpha ordered. Also should add YUIdoc documentation.

JoshSmith replied:
Documentation is for the internals, I mean.

JoshSmith commented:
Still in progress, right?

JoshSmith commented:
I've moved everything to `project/settings/donations` and rewrote some of the logic for when you can add/edit/cancel.

Also added information to the top of the page.

Removed the `currency` helper and replaced with `format-currency`; `currency` was fairly broken. Changed some logic in `format-currency`.

Nothing I've changed has tests for it. Not completely finished with the work here.

begedin commented:
@JoshSmith I was unable to completely finalize this, but I got close.

*I fixed the tests that got broken
*I added a project ability, with the option to call `can('manage donation goals in project)`, which returns true if user is the owner of the project's organization. The current credentials of "at least admin" for `can('manage organization')` were not enough, since we explicitly said API-side that only owners can do this.
* I updated the tests to match that a well. I also added an authentication helper which enables quickly creating an organization member of a specific role in an organization and authenticating with that member.
* I further updated `format-currency` so it works with an integer amount in cents, since that was my understanding we were going to use, API-side. If that's wrong, then I really think we should either rename `DonationGoal.amount` or `StripePlan.amount`, since having them be named the same and one be in dollars while the other is in cents is confusing.
* [ ] We still need tests for the project ability.

sbatson5 commented on app/components/donation-goal-edit.js:
```diff
@@ -0,0 +1,85 @@
+     * @method cancel
+     */
+    cancel() {
+      this.sendAction('cancel');
```
We seem to be using closure actions with these components:
https://github.com/code-corps/code-corps-ember/pull/634/files#diff-e54aed91f3708411eb6f637d00ab5400R15

Thoughts on `this.get('cancel')()`?

begedin replied:
I'm not sure I understand, so any advice is welcome. What's the advantage?

begedin replied:
I had a superficial understanding of it, so I made an effort to comprehend it a bit better. This video was the most helpful for me - https://www.emberscreencasts.com/posts/76-closure-actions-in-ember-20-components-passing-functions

I made use of closures and currying to the point where I now pass in the functions from controller level (closures are sadly not possible with route actions), which cleared a lot of "propagate this action up" code and made templates more explicit and easier to understand, in my opinion.

sbatson5 replied:
For route actions, we actually use `ember-route-action-helper` which allows us to pass actions from our route down to components https://github.com/code-corps/code-corps-ember/blob/develop/package.json#L78

You can do `cancel=(route-action "cancel")`.  It has been discussed that route-actions as closure actions should make it into an Ember release at some point.

sbatson5 replied:
Also, regarding that screencast, it is slightly outdated. It is not encouraged to call the actions from the `attrs` directly.  Locks has a good article about why you shouldn't call attrs directly: https://locks.svbtle.com/to-attrs-or-not-to-attrs

Basically, attrs is a Glimmer Components thing, which hasn't landed in a stable version of Ember quite yet.  In Classic Components (curly brace components), attrs have a slightly different meaning behind the scenes.

JoshSmith commented:
@sbatson5 could use your opinion on `format-currency`. I like what you had previously and think it's okay to be dealing in either dollars or cents when just inputting data into a form field, but that we should do our conversion to cents just when we `POST` to the server. What do you think?

sbatson5 commented:
@JoshSmith I don't know if I have a strong opinion on when the conversion should happen. When I wrote the helper, it made sense because it was coming from user input. Do we want to make that conversion when the request goes out or do we want the API to handle it?

What if we created a transform on the Ember side that took care of the serializing and deserializing for us?  That we can honor the user input in these helpers and that would handle the conversion to and from the API.

I can PR something real quick to see how we like it.

begedin commented:
@JoshSmith If we merge #654, then we will still need the `format-currency` helper, but we will have to revert it back to what you wrote. Just to make that clear. I do think we probably need both either way, though, since we're dealing with user input in the `donation-goal-edit` component, so the flow here should be:

* [ ] Merge #654 
* [ ] Add it to the `donation-goal-edit` component here.
* [ ] Update tests

begedin commented:
I added the missing unit tests. While doing this, I took a look at our `has-attributes` helper for model tests. I took a look at how `testBelongsTo` and `testHasMany` work and I repurposed `has-attributes` into a `testAttributes` helper, which works in a similar way.

I believe that, in a separate issue, we should get rid of `has-attributes` and move it's behavior into `testAttributes`, since it's less hacky that way (it would not override a `QUnit` prototype.

Additionally, we can then merge `helpers/relationships` and the now added `helpers/attributes` into `helpers/model.js`, for IMO slightly better code organization.

@JoshSmith Let me know if I should create an issue for that.

You can take a look at how the new stuff works in `tests/unit/models/donation-goal`

begedin commented:
This should finally be ready for final review and (hopefully) merge.

begedin commented on mirage/config.js:
```diff
@@ -136,6 +136,15 @@ export default function() {
+  this.get('/donation-goals', { coalesce: true });
+  this.get('/donation-goals/:id');
+  this.patch('/donation-goals/:id');
+  this.post('/donation-goals');
```
I really don't think we need to comment on each and every mirage config shorthand. A comment like

```Javascript
// POST /categories
this.post('/categories');
```

really just tells us what the code is already telling us.

begedin commented on mirage/factories/donation-goal.js:
```diff
@@ -0,0 +1,6 @@
+import { Factory, faker } from 'ember-cli-mirage';
+
+export default Factory.extend({
+  amount: faker.list.random(10000, 20000, 15020),
```
Mirage does not support custom attribute transforms, as far as I can tell, so we have to deal with cents. It makes sense, really. Mirage is there to mock the API, and the API deals with cents.

begedin commented on tests/helpers/attributes.js:
```diff
@@ -0,0 +1,17 @@
+
+    assert.hasAttributes(actualAttributes, expectedAttributes);
+  });
+}
```
We can move this and `tests/helpers/relationhips` into a common `test/helpers/model` and then get rid of `has-attributes`, since that part can be mimicked from within this helper, without the need to monkeypatch the `QUnit` prototype.

sbatson5 commented on app/controllers/project/settings/donations.js:
```diff
@@ -0,0 +1,67 @@
+import Ember from 'ember';
+
+const { Controller } = Ember;
```
You should be able to use the `ember-route-action-helper` if you don't want to go through the controller.  Although, I don't see anything wrong with using a controller in this case.
